### PR TITLE
Add marl::containers::list, use for marl::ConditionVariable

### PR DESCRIPTION
**Pull request by ben-clayton**
Provides substantial performance improvements.

### CODE REVIEW

File: include/marl/containers.h [hunk 1]
```diff
@@ -243,6 +242,201 @@ void vector<T, BASE_CAPACITY>::free() {
...
+  inline iterator emplace_front(U&&);
+  inline void erase(iterator);
+
+ private:
+  void grow(size_t count);
+
+  static void unlink(Entry* entry, Entry*& list);
+  static void link(Entry* entry, Entry*& list);
+
+  Allocator* const allocator;
```
Comment (ben-clayton) [inline include/marl/containers.h]: TODO: Need to deal with copy and move constructors / operators. May just delete.

File: include/marl/memory.h [hunk 2]
```diff
@@ -38,7 +38,8 @@ struct Allocation {
     Undefined = 0,
     Stack,   // Fiber stack
     Create,  // Allocator::create(), make_unique(), make_shared()
-    Vector,  // marl::vector<T>
+    Vector,  // marl::containers::vector<T>
+    List,  // marl::containers::vector<T>
```
Comment (ben-clayton) [inline include/marl/memory.h]: TODO: clang-format

File: include/marl/containers.h (hunk 1 above)
Comment (ben-clayton) [reply include/marl/containers.h]: Done

File: include/marl/memory.h (hunk 2 above)
Comment (ben-clayton) [reply include/marl/memory.h]: Done

**Comment by ben-clayton** [comment]
Benchmarks show performance improvements ~ 1.4x for certain tests (mostly `Schedule/Event`).

File: include/marl/containers.h [hunk 3]
```diff
@@ -243,6 +242,207 @@ void vector<T, BASE_CAPACITY>::free() {
...
+}
+
+template <typename T>
+size_t list<T>::size() const {
+  return size_;
+}
+
+template <typename T>
+template <typename U>
+typename list<T>::iterator list<T>::emplace_front(U&& el) {
```
Comment (AWoloszyn) [inline include/marl/containers.h]: emplace_front is usually

template<typename T> template<typename... Args>
typename list<T>::iterator list<T>::emplace_front(Args&&... args) {
   ....
   new (&entry->data) T(std::forward<T>(args)...);
}

**Comment by AWoloszyn** [comment]
Looks good, one simple comment about interop with std::list.

File: include/marl/containers.h (hunk 3 above)
Comment (ben-clayton) [reply include/marl/containers.h]: Done. Thanks.

**Pull request merged**